src/components/cc-domain-management/cc-domain-management.js

```diff
@@ -0,0 +1,819 @@
+    if (error === 'empty') {
+      return i18n('cc-domain-management.form.domain.error.empty');
+    }
+    if (error === 'invalid-format') {
+      return i18n('cc-domain-management.form.domain.error.format');
+    }
+    if (error === 'invalid-wildcard') {
+      return i18n('cc-domain-management.form.domain.error.wildcard');
+    }
+    if (error?.code != null && error?.code === 'hostname-contains-path') {
+      return i18n('cc-domain-management.form.domain.error.contains-path', { path: error.pathWithinHostname });
+    }
+    return null;
+  }
+
+  /** @param {ClipboardEvent} event */
+  _onDomainPaste (event) {
+    if (this.domainFormState.type !== 'idle' || this.domainFormState.hostname.value.trim() !== '') {
+      return;
+    }
+
+    event.preventDefault();
+
+    const clipboardValue = event.clipboardData.getData('text');
+
+    try {
+      const { hostname, pathname, isWildcard } = parseDomain(clipboardValue);
+
+      this.domainFormState = {
+        ...this.domainFormState,
+        hostname: {
+          ...this.domainFormState.hostname,
+          value: getHostWithWildcard(hostname, isWildcard),
+        },
+        pathPrefix: {
+          ...this.domainFormState.pathPrefix,
+          value: pathname.replace(/\/$/, ''),
+        },
+      };
+    }
+    catch {
+      this.domainFormState = {
+        ...this.domainFormState,
+        hostname: {
+          ...this.domainFormState.hostname,
+          value: clipboardValue,
+        },
+      };
+    }
+
+  }
+
+  /** @param {SubmitEvent} e */
+  _onDomainSubmit (e) {
+    e.preventDefault();
+
+    if (this.domainFormState.type !== 'idle') {
+      return;
+    }
+
+    // we trim and we'll send this value to the API but we don't change what is displayed within the input
+    const hostnameValue = this.domainFormState.hostname.value.trim();
+    const pathPrefixValue = this.domainFormState.pathPrefix.value.trim();
+    const error = this._getErrorCode(hostnameValue);
+
+    if (error != null) {
+      this.domainFormState = {
+        ...this.domainFormState,
+        hostname: {
+          ...this.domainFormState.hostname,
+          error,
+        },
+      };
+
+      this._domainInputRef.value.focus();
+      return;
+    }
+
+    this.domainFormState = {
+      ...this.domainFormState,
+      hostname: {
+        value: this.domainFormState.hostname.value,
+      },
+    };
+
+    // we do this to strip off unwanted parts like query parameters for instance
+    const { hostname, pathname, isWildcard } = parseDomain(hostnameValue + pathPrefixValue);
+
+    dispatchCustomEvent(this, 'add', {
+      hostname,
+      pathPrefix: pathname,
+      isWildcard,
+    });
+  }
+
+  /** @param {FormattedDomainInfo} domainInfo */
+  _onMarkPrimary (domainInfo) {
+    const { id, hostname, pathPrefix, isWildcard, isPrimary } = domainInfo;
+    return () => dispatchCustomEvent(this, 'mark-as-primary', { id, hostname, pathPrefix, isWildcard, isPrimary });
```

Comment:
question: why do we need to return an arrow function here? :thinking:

Reply:
This is an equivalent pattern for this case
```tsx
          <cc-button
            class="delete-domain"
            ...
            @cc-button:click=${() => this._onDelete(domainInfo)}
          >
            ${i18n('cc-domain-management.list.btn.delete.text')}
          </cc-button>
```

You want to be able to pass arguments from the render to the callback listener.

For this case I could have sticked with the pattern above but when you also want to pass the event, TypeScript will complain about the event type and you end up with this:

```tsx
          <cc-button
            class="delete-domain"
            ...
            @cc-button:click=${
              /** @param {Event} */
              (event) => this._onDelete(event, domainInfo)}
            }
          >
            ${i18n('cc-domain-management.list.btn.delete.text')}
          </cc-button>
```

Since we don't like that we discussed it with Hubert and he suggested this solution instead. This pattern is already used here if you want another example: https://github.com/CleverCloud/clever-components/blob/master/src/components/cc-addon-backups/cc-addon-backups.js#L233

We can discuss it in sync if it's unclear and we should probably discuss it in team anyway :man_shrugging:

Reply:
I may be wrong but I think there's a difference between those two implementations:
- when `_onMarkPrimary()` returns an arrow function, the `const { ... } = domainInfo;` line is executed at runtime, when the component is displayed,
- when `_onMarkPrimary()` returns nothing and simply dispatch the event, the `const { ... } = domainInfo;` line is executed when the click occurs.

It's because the HTML template differs as well:
- with the arrow function: `@cc-button:click=${this._onMarkPrimary(domainInfo)}`,
- without the arrow function: `@cc-button:click=${(event) => this._onMarkPrimary(event, domainInfo)}`.

A caveat with the arrow function is that, if the data tied to the DOM element (here, it's `domainInfo`) is updated dynamically from its initial value during the component lifetime, the data fired alongside the event will mismatch (I'm not sure though, should be tested).

Reply:
This is something I wanted to test as well when we initially used it within the `cc-addon-backups` but I didn't have time + I felt the only case where it might happen was if the piece of data we wanted to send was not part of a reactive property which is very unlikely :thinking: